sktime/utils/numba/stats.py

```diff
@@ -11,9 +11,27 @@ def mean(X):
     return np.mean(X)
 
 
-@njit(fastmath=True, cache=True)
 def median(X):
-    """Numba median function for a single time series."""
+    """Numba median function for a single time series.
+
+    This function clones the data if the array passed is not writable before finding
+    the median. This is because of a known bug in the interaction between numba and
+    numpy np.median function that causes an error with threaded jobs with large
+    datasets. See
+    https://github.com/numba/numba/issues/3070
+    and
+    https://github.com/sktime/sktime/issues/3582
+    """
+    if X.flags.writeable:
+        return _median(X)
+    else:
+        X = X.copy()
+        X.setflags(write=True)
+        return _median(X)
+
+
+@njit(fastmath=True, cache=True)
+def _median(X):
     return np.median(X)
```

Comment:
Is calling np.median() from within @njit faster than calling np.median directly?

Reply:
> Sorry for asking the obvious? Is calling numpy methods from within numba `@njit` faster than calling numpy directly?

I profiled both options, and the time is about the same. However, if we do not njit this function, then the problem will just occur elsewhere when np.median is called in a njit'ed method, thus shifting the problem elsewhere. @chrisholder and @MatthewMiddlehurst may want to comment

Reply:
Ok, but in this case there should be an njit tag in line 14?

Reply:
see above :) I am very much black boxing this, its all about the interaction with the Parallel threading structure of caching. There may be a neater solution here, I don't like the cloning, but this has the benefit of fixing the problem

Reply:
I am a bit confused :)

As I understand it: To use median() from an njit'ed method, all sub-methods must be njit'ed, too.

As  `median(X)` is not njit'ed, we must not call this from an njit'ed method. But we cannot call `_median(X)` directly from an njit'ed method either, as it has the readonly problem?

Reply:
I dont think you can actually call the new function in other njit functions, which may be a bit of an oversight now that i think about it 🙂

Reply:
> see above :) I am very much black boxing this, its all about the interaction with the Parallel threading structure of caching. There may be a neater solution here, I don't like the cloning, but this has the benefit of fixing the problem

Which method is calling median? Is this method njit'ed at all? E.g. does it need an njit'ed median at all?

Reply:
This is less about whether it is faster, and more about requiring a numba median function for other numba functions. Just removing the njit is kicking the can down the road to the next numba function that requires the median.

Reply:
>  a numba median function for other numba functions

Sure, but then all methods should be njit'ed? Or am I missing on something?

Reply:
hmmm, I'm confused too need some input from @chrisholder maybe? The basic problem is we cannot call np.median directly from any njit'ed method without allowing this bug, so this non-njit wrapper is meant to mitigate that. It definitely runs, tried it on the two cases which previously crashed (mine and the data from =3582). Perhaps we are simply un-njit-ing the whole thing by doing it like this?  

I think the best solution may just be to implement quickselect (or whatever the best median finder is) in numba

Reply:
yeah

Reply:
Would "@jit" work at median() ? Not @njit

Reply:
> I dont think you can actually call the new function in other njit functions, which may be a bit of an oversight now that i think about it 🙂

ok, so it seems its only used here 
![image](https://user-images.githubusercontent.com/9594042/196400928-82a33afa-646a-4dc1-88c5-b3d2f71e4425.png)

@MatthewMiddlehurst maybe just refactor it out and abandon this method?

Reply:
and use np.median directly?

Reply:
Sounds ok to me.

`numpy.partition` could be a solution to not using median, too. It is based on quicksort, and can be used to return the median, too by passing the index of the median to it.

Reply:
Can just delete the functions and replace the usage for now as you say